### PR TITLE
VACMS-12619: Add field validation facility health services

### LINKED ISSUE #12619: CMS: Adjust field validation/reset on VAMC Facility health service
## Description
See [this comment](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/12232#issuecomment-1431464681).

Field selections do not get reset when changing the Section field, so it's possible to end up with weird field combinations:
image

Further, these inconsistent field values are not validated on save.

So, fields should be reset and/or validated on save.

## Acceptance Criteria
Field Reset 
- [ ] (If applicable) Node create: Two dependent fields are reset when Section field is changed. (This already happens; ensure no regression).
- [ ] (If applicable) Node edit: Two dependent fields are reset when Section field is changed. ([There's an issue with this.](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/12619#issuecomment-1440062908))

Field Validation
- [ ] When the Facility and VAMC system health service match the chosen Section, the node is saved. 
- [ ] When the Facility does not match the section:
  - [ ] The node does not save.
  - [ ] A message is shown to the editor: "The Facility does not match the Section. Please select a matching option."
- [ ] When the VAMC system health service does not match the Section:
  - [ ] The node does not save.
  - [ ] A message is shown to the editor: "The VAMC system health service does not match the Section. Please select a matching option."
- [ ] When both the Facility and the VAMC system health service do not match the Section:
  - [ ] The node does not save.
  - [ ] A message is shown to the editor: "The Facility and VAMC system health service do not match the section. Please select matching options."

### Team
Please check the team(s) that will do this work.

- [ ] `CMS Team`
- [ ] `Public Websites`
- [x] `Facilities`
- [ ] `User support`

### PR DESCRIPTION
**Pull request by omahane**
## Description

Closes #12619

## Testing done
Cypress
Manual

## Screenshots
Screen Shot 2023-03-06 at 4 17 53 PM
Screen Shot 2023-03-06 at 4 16 18 PM
Screen Shot 2023-03-06 at 4 16 59 PM

### Edit a VAMC facility health service
- [x] Log in as Ryan.Stubblebine@va.gov
- [x] Edit [My HealtheVet coordinator - Belmont County VA Clinic](https://pr12865-anjh0saa6ev6qhdgsr9wifu9sftbjnip.ci.cms.va.gov/node/2333/edit)
- [x] Change the Section to "---VA Lebanon health care" 
- [x] Click "Save draft and continue editing"
- [x] Validate that the Facility field presents a relevant error message
- [x] Validate that the VAMC system health service field presents a relevant error message
- [x] Change the Facility to "Berks County VA Clinic | VA Lebanon health care"
- [x] Click "Save draft and continue editing"
- [x] Validate that the VAMC system health service field presents a relevant error message
- [x] Change the Section to "---VA Pittsburgh health care"
- [x] Click "Save draft and continue editing"
- [x] Validate that the Facility field presents a relevant error message
- [x] Change the Facility to "Fayette County VA Clinic | VA Pittsburg health care"
- [x] Click "Save draft and continue editing"
- [x] Validate that the service is changed.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `CMS Team`
- [ ] `Public websites`
- [x] `Facilities`
- [ ] `User support`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CODE REVIEW

File: docroot/modules/custom/va_gov_vamc/va_gov_vamc.module [hunk 1]
```diff
@@ -187,3 +188,27 @@ function va_gov_vamc_inline_entity_form_entity_form_alter(&$entity_form, FormSta
...
+function va_gov_vamc_entity_bundle_field_info_alter(&$fields, EntityTypeInterface $entity_type, $bundle) {
+  $list_item_bundles = [
+    'health_care_local_health_service',
+  ];
+  if (($entity_type->id() === 'node')
+  && (in_array($bundle, $list_item_bundles))) {
+    // Focused on VAMC facility health services.
+    if (isset($fields['field_administration'])
+    && isset($fields['field_service_location'])) {
+      // Validate the Facility selection based on Section.
```
Comment (JayDarnell) [inline docroot/modules/custom/va_gov_vamc/va_gov_vamc.module:204]: Section probably doesn't need to be capitalized here but that isn't a show stopper by any means. I do acknowledge that you are referencing a field label so showing it as capitalized does make sense.

**Review by JayDarnell — DISMISSED** [review]
Code looks good and QA steps pass.

File: docroot/modules/custom/va_gov_vamc/va_gov_vamc.module (hunk 1 above)
Comment (omahane) [reply docroot/modules/custom/va_gov_vamc/va_gov_vamc.module:204]: Yes, that is what I was thinking. Should I put it in quotes?

Comment (JayDarnell) [reply docroot/modules/custom/va_gov_vamc/va_gov_vamc.module:204]: I'm not sure you even need to change it to be honest. I just noticed that the field labels in the comments for the va_gov_lovell module weren't capitalized when I compared the two. Consistency is great but I don't know if its really that important. I would be fine to merge it as is but if you feel compelled to align them go for it.

File: docroot/modules/custom/va_gov_vamc/src/Plugin/Validation/Constraint/SectionMatcherValidator.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  public function validate($items, Constraint $constraint) {
+    $entity = $items->getEntity();
+    $sectionTermID = $entity->field_administration->target_id;
+    $fieldLabel = $items->getFieldDefinition()->getLabel();
+    $nodeStorage = \Drupal::entityTypeManager()->getStorage('node');
+    $termStorage = \Drupal::entityTypeManager()->getStorage('taxonomy_term');
+    $sectionName = $termStorage->load($sectionTermID)->getName();
+    foreach ($items as $item) {
+      /** @var \Drupal\va_gov_vamc\Plugin\Validation\Constraint\SectionMatcher $constraint */
+      $refEnt = $nodeStorage->load($item->target_id);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_vamc/src/Plugin/Validation/Constraint/SectionMatcherValidator.php:25]: The is a short enough abbreviation / truncation that I am not sure what it is.  A more meaningful name would help convey that more reliably.

**Review by swirtSJW — COMMENTED** [review]
Overall looks great.  Just a question and a suggestion.

File: docroot/modules/custom/va_gov_vamc/src/Plugin/Validation/Constraint/SectionMatcherValidator.php (hunk 2 above)
Comment (omahane) [reply docroot/modules/custom/va_gov_vamc/src/Plugin/Validation/Constraint/SectionMatcherValidator.php]: I have expanded this variable.

File: docroot/modules/custom/va_gov_vamc/src/Plugin/Validation/Constraint/SectionMatcherValidator.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+   */
+  public function validate($items, Constraint $constraint) {
+    $entity = $items->getEntity();
+    $sectionTermID = $entity->field_administration->target_id;
+    $fieldLabel = $items->getFieldDefinition()->getLabel();
+    $nodeStorage = \Drupal::entityTypeManager()->getStorage('node');
+    $termStorage = \Drupal::entityTypeManager()->getStorage('taxonomy_term');
+    $sectionName = $termStorage->load($sectionTermID)->getName();
+    foreach ($items as $item) {
+      /** @var \Drupal\va_gov_vamc\Plugin\Validation\Constraint\SectionMatcher $constraint */
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_vamc/src/Plugin/Validation/Constraint/SectionMatcherValidator.php]: I have moved the type declaration down to where $constraint is used. @JayDarnell had said that he was getting a PHPStan error without it.

**Review by swirtSJW — APPROVED** [review]
Nice work Omahane.

**Comment by omahane** [comment]
Approved and ready to merge (after updating with base branch).
Screen Shot 2023-03-09 at 10 50 37 PM

**Pull request merged**